src/arch/arm/64/kernel/vspace.c

```diff
@@ -233,68 +233,173 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+
+            armKSGlobalKernelPDs[pud_idx][pd_idx]
+                = pte_pte_page_new(
 #ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-    /* verify that the kernel window as at the second entry of the PGD */
-    assert(GET_KPT_INDEX(PPTR_BASE, KLVL_FRM_ARM_PT_LVL(0)) == 1);
+                      /* XN */ 0,
 #else
-    /* verify that the kernel window as at the last entry of the PGD */
-    assert(GET_KPT_INDEX(PPTR_BASE, KLVL_FRM_ARM_PT_LVL(0)) == BIT(PT_INDEX_BITS) - 1);
+                      /* UXN */ 1,
 #endif
-    assert(IS_ALIGNED(PPTR_BASE, seL4_LargePageBits));
-    /* verify that the kernel device window is 1gb aligned and 1gb in size */
-    assert(GET_KPT_INDEX(PPTR_TOP, KLVL_FRM_ARM_PT_LVL(1)) == BIT(PT_INDEX_BITS) - 1);
-    assert(IS_ALIGNED(PPTR_TOP, seL4_HugePageBits));
-
-    /* place the PUD into the PGD */
-    armKSGlobalKernelPGD[GET_KPT_INDEX(PPTR_BASE, KLVL_FRM_ARM_PT_LVL(0))] = pte_pte_table_new(
-                                                                                 addrFromKPPtr(armKSGlobalKernelPUD));
-
-    /* place all PDs except the last one in PUD */
-    for (idx = GET_KPT_INDEX(PPTR_BASE, KLVL_FRM_ARM_PT_LVL(1)); idx < GET_KPT_INDEX(PPTR_TOP, KLVL_FRM_ARM_PT_LVL(1));
-         idx++) {
-        armKSGlobalKernelPUD[idx] = pte_pte_table_new(
-                                        addrFromKPPtr(&armKSGlobalKernelPDs[idx][0])
-                                    );
+                      /* page_base_address */ paddr,
+                      /* global (nG) */ 0,
+                      /* access flag (AF) */ 1,
+                      /* shareability (SH) */ SMP_TERNARY(SMP_SHARE, 0),
+                      /* AP */ APFromVMRights(VMKernelOnly),
+                      /* AttrIndx */ NORMAL
+                  );
+
+            assert(GET_KPT_INDEX(vaddr, KLVL_FRM_ARM_PT_LVL(1)) == pud_idx);
+            assert(GET_KPT_INDEX(vaddr, KLVL_FRM_ARM_PT_LVL(2)) == pd_idx);
+            paddr += BIT(seL4_LargePageBits);
+            vaddr += BIT(seL4_LargePageBits);
+        }
     }
+}
 
-    /* map the kernel window using large pages */
-    vaddr = PPTR_BASE;
-    for (paddr = PADDR_BASE; paddr < PADDR_TOP; paddr += BIT(seL4_LargePageBits)) {
-        armKSGlobalKernelPDs[GET_KPT_INDEX(vaddr, KLVL_FRM_ARM_PT_LVL(1))][GET_KPT_INDEX(vaddr,
-                                                                                         KLVL_FRM_ARM_PT_LVL(2))] = pte_pte_page_new(
+BOOT_CODE void map_kernel_elf_image(const word_t kernel_mappings_pud_idx)
+{
+    compile_assert(elf_paddr_aligned, IS_ALIGNED(KERNEL_ELF_PADDR_BASE_RAW, seL4_LargePageBits));
+    compile_assert(elf_base_aligned, IS_ALIGNED(KERNEL_ELF_BASE, seL4_LargePageBits));
+    /* asserts because ELF_TOP is dynamic */
+    assert(GET_KPT_INDEX(KERNEL_ELF_BASE, KLVL_FRM_ARM_PT_LVL(1)) == kernel_mappings_pud_idx);
+    assert(GET_KPT_INDEX(KERNEL_ELF_TOP, KLVL_FRM_ARM_PT_LVL(1)) == kernel_mappings_pud_idx);
```

Comment:
Considering you know what it should be, wouldn't it make more sense to move the `armKSGlobalKernelPUD[kernel_mappings_pud_idx] = pte_pte_table_new(addrFromKPPtr(&armKSGlobalKernelPDs[kernel_mappings_pud_idx][0]));` here and remove the function argument?